crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,124 @@
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send + 'static> Global<T> {
+    /// Gets the current value of the global. Returns `None` if the global has not
+    /// yet been initialized.
+    pub fn get<'cx, 'a, C>(&self, cx: &'a mut C) -> Option<&'cx T>
+    where
+        C: Context<'cx>,
+    {
+        // Safety: The type bound Global<T> and the fact that every Global has a unique
+        // id guarantees that the cell is only ever assigned instances of type T.
+        let r: Option<&T> =
+            GlobalCell::get(cx, self.id()).map(|value| value.downcast_ref().unwrap());
+
+        // Safety: Since the Box is immutable and heap-allocated, it's guaranteed not to
+        // move or change for the duration of the context.
+        unsafe { std::mem::transmute::<Option<&'a T>, Option<&'cx T>>(r) }
+    }
+
+    /// Gets the current value of the global, initializing it with `value` if it has
+    /// not yet been initialized.
+    pub fn get_or_init<'cx, 'a, C>(&self, cx: &'a mut C, value: T) -> &'cx T
+    where
+        C: Context<'cx>,
+    {
+        // Safety: The type bound Global<T> and the fact that every Global has a unique
+        // id guarantees that the cell is only ever assigned instances of type T.
+        let r: &T = GlobalCell::get_or_init(cx, self.id(), Box::new(value))
+            .downcast_ref()
+            .unwrap();
+
+        // Safety: Since the Box is immutable and heap-allocated, it's guaranteed not to
+        // move or change for the duration of the context.
+        unsafe { std::mem::transmute::<&'a T, &'cx T>(r) }
+    }
+
+    /// Gets the current value of the global, initializing it with the result of
+    /// calling `f` if it has not yet been initialized.
+    pub fn get_or_init_with<'cx, 'a, C, F>(&self, cx: &'a mut C, f: F) -> &'cx T
+    where
+        C: Context<'cx>,
+        F: FnOnce() -> T,
+    {
+        // Safety: The type bound Global<T> and the fact that every Global has a unique
+        // id guarantees that the cell is only ever assigned instances of type T.
+        let r: &T = GlobalCell::get_or_init_with(cx, self.id(), || Box::new(f()))
+            .downcast_ref()
+            .unwrap();
+
+        // Safety: Since the Box is immutable and heap-allocated, it's guaranteed not to
+        // move or change for the duration of the context.
+        unsafe { std::mem::transmute::<&'a T, &'cx T>(r) }
+    }
+
+    /// Gets the current value of the global, initializing it with the result of
+    /// calling `f` if it has not yet been initialized. Returns `Err` if the
+    /// callback triggers a JavaScript exception.
+    ///
+    /// During the execution of `f`, calling any methods on this `Global` that
+    /// attempt to initialize it will panic.
+    pub fn get_or_try_init<'cx, 'a, C, F>(&self, cx: &'a mut C, f: F) -> NeonResult<&'cx T>
```

Comment:
Does this need to be constrained to a `NeonResult`? Since we are panicking instead of throwing an exception, I think this can be more generally `Result<&'cx T, E>`.

This would make it a bit more ergonomic for cases that are fallible for things that *aren't* Neon related.

```rust
let runtime = RUNTIME.get_or_try_init(&mut cx, |_| Runtime::new())?;
```

Reply:
That's a really cool idea!

Reply:
Implemented and pushed